core/src/main/java/hivemall/smile/classification/MaxEntUDTF.java

```diff
@@ -0,0 +1,440 @@
+
+        if (!labels.isEmpty()) {
+            Matrix x = matrixBuilder.buildMatrix();
+            this.matrixBuilder = null;
+            int[] y = labels.toArray();
+            this.labels = null;
+
+            // run training
+            train(x, y);
+        }
+
+        // clean up
+        this.featureListOI = null;
+        this.featureElemOI = null;
+        this.labelOI = null;
+    }
+    
+    private void checkOptions() throws HiveException {
+    	if (_USE_SMOOTHING == false && _SMOOTHING_OBSERVATION != 0.1) {
+            throw new HiveException("Instructions received to avoid smoothing, but smoothing constant is set [" + _SMOOTHING_OBSERVATION + "]");
+        }
+    }
+    
+    /**
+     * @param x features
+     * @param y label
+     * @param attrs attribute types
+     * @param numTrees The number of trees
+     * @param numVars The number of variables to pick up in each node.
+     * @param seed The seed number for Random Forest
+     */
+    private void train(@Nonnull Matrix x, @Nonnull final int[] y) throws HiveException {
+        final int numExamples = x.numRows();
+        if (numExamples != y.length) {
+            throw new HiveException(String.format("The sizes of X and Y don't match: %d != %d",
+                numExamples, y.length));
+        }
+        checkOptions();
+
+        int[] labels = SmileExtUtils.classLables(y);
+        Attribute[] attributes = SmileExtUtils.attributeTypes(_attributes, x);
+
+        if (logger.isInfoEnabled()) {
+            logger.info("real: " + _real + ", smoothing: " + this._USE_SMOOTHING + ", smoothing constant: "
+                    + _SMOOTHING_OBSERVATION);
+        }
+
+        IntMatrix prediction = new DoKIntMatrix(numExamples, labels.length); // placeholder for out-of-bag prediction
+        AtomicInteger remainingTasks = new AtomicInteger(_numTrees);
+        List<TrainingTask> tasks = new ArrayList<TrainingTask>();
+        for (int i = 0; i < _numTrees; i++) {
+            tasks.add(new TrainingTask(this, i, attributes, x, y, prediction, remainingTasks));
+        }
+
+        MapredContext mapredContext = MapredContextAccessor.get();
+        final SmileTaskExecutor executor = new SmileTaskExecutor(mapredContext);
+        try {
+            executor.run(tasks);
+        } catch (Exception ex) {
+            throw new HiveException(ex);
+        } finally {
+            executor.shotdown();
+        }
+    }
+    
+    /**
+     * Synchronized because {@link #forward(Object)} should be called from a single thread.
+     * 
+     * @param accuracy
+     */
+    synchronized void forward(final int taskId, @Nonnull final Text model,
+    		@Nonnull Attribute[] attributes,
+            @Nonnegative final double accuracy, final int[] y,
+            @Nonnull final IntMatrix prediction, final boolean lastTask) throws HiveException {
+        int oobErrors = 0;
+        int oobTests = 0;
+        if (lastTask) {
+            // out-of-bag error estimate
```

Comment:
Out-Of-Bag testing is a method for RandomForest. Should not use training examples used in the input of training.
https://www.quora.com/What-is-the-out-of-bag-error-in-Random-Forests

OOB testing is not needed for MaxEnt classifier in my understanding.